core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -410,15 +411,39 @@ abstract class AbstractFetcherThread(name: String,
     } finally partitionMapLock.unlock()
   }
 
-  private def markPartitionFailed(topicPartition: TopicPartition): Unit = {
+  private def markPartitionFailed(topicPartition: TopicPartition, exception: Throwable): Unit = {
     partitionMapLock.lock()
     try {
-      failedPartitions.add(topicPartition)
+      failedPartitions.add(topicPartition, exception)
       removePartitions(Set(topicPartition))
     } finally partitionMapLock.unlock()
     warn(s"Partition $topicPartition marked as failed")
   }
 
+  /**
+   * update the epoch of pending partition. If the partition was failed due to fenced error, the partition is added
+   * back to pending pool with new epoch and offset.
+   * @param partitionAndOffsets partitions to update
+   */
+  def updateEpochs(partitionAndOffsets: Map[TopicPartition, InitialFetchState]): Unit = {
+    partitionMapLock.lockInterruptibly()
+    try {
+      partitionAndOffsets.foreach {
+        case (tp, state) =>
+          failedPartitions.exception(tp) match {
+            case Some(e) if e.getClass == Errors.FENCED_LEADER_EPOCH.exception().getClass =>
```

Comment:
nit: Can this be handled with a nested match/case using types? (rather than using `getClass`)

Reply:
not sure how to match dynamical class type in scala. But I rewrite this match pattern by ```Errors.forException``` which converts the exception to Error.
```scala
          failedPartitions.exception(tp) match {
            case Some(e) =>
              Errors.forException(e) match {
                case Errors.FENCED_LEADER_EPOCH =>
                  addPartitions(Map(tp -> OffsetAndEpoch(state.initOffset, state.currentLeaderEpoch)))
                  failedPartitions.removeAll(Set(tp))
                case _ =>
                  // the exception is NOT caused by inconsistent epoch so we can't resume the error
              }
            case None =>
              Option(partitionStates.stateValue(tp)).map(_.copy(currentLeaderEpoch = state.currentLeaderEpoch))
                .foreach(partitionStates.updateAndMoveToEnd(tp, _))
          }
```